ml-agents/mlagents/trainers/optimizer/torch_optimizer.py

```diff
@@ -49,25 +55,115 @@ def create_reward_signals(self, reward_signal_configs):
                 reward_signal, self.policy.behavior_spec, settings
             )
 
+    def _evaluate_by_sequence(
+        self, tensor_obs: List[torch.Tensor], initial_memory: np.ndarray
+    ) -> Tuple[Dict[str, torch.Tensor], AgentBufferField, torch.Tensor]:
+        """
+        Evaluate the batch sequence-by-sequence, assembling the result. This enables us to get the
+        intermediate memories for the critic.
+        """
+        num_experiences = tensor_obs[0].shape[0]
+        all_next_memories = AgentBufferField()
+        # In the buffer, the 1st sequence are the ones that are padded. So if seq_len = 3 and
+        # trajectory is of length 10, the 1st sequence is [pad,pad,obs].
+        # Compute the number of elements in this padded seq.
+        leftover = num_experiences % self.policy.sequence_length
+
+        # Compute values for the potentially truncated initial sequence
+        _mem = initial_memory
+        seq_obs = []
+
+        first_seq_len = self.policy.sequence_length
```

Comment:
Any reason for padding the end of the first sequence and not the last sequence ?

Reply:
Neither is padded in this function, but the buffer pads the front of the first sequence, so this function accounts for that. I left the padding in the buffer as-is and still need to experiment with that change; it will likely be a future PR.

Reply:
Then why is the first sequence handled differently ?